folsom/src/test/java/com/spotify/folsom/client/TimeoutCheckerTest.java

```diff
@@ -18,33 +18,72 @@
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertTrue;
 
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
 import java.util.concurrent.TimeUnit;
 import org.junit.Test;
 
 public class TimeoutCheckerTest {
 
-  final TimeoutChecker<String> sut = TimeoutChecker.create(TimeUnit.MILLISECONDS, 100);
+  private TimeoutChecker sut = TimeoutChecker.create(TimeUnit.MILLISECONDS, 100);
+  private Request<Integer> request;
 
   @Test
-  public void testTimeout() throws Exception {
-    assertFalse(sut.check("foo"));
-    assertFalse(sut.check("foo"));
-    Thread.sleep(200);
-    assertTrue(sut.check("foo"));
+  public void testTimeout0ms() throws Exception {
+    request = new DummyRequest();
+    assertFalse(sut.check(request));
   }
 
   @Test
-  public void testNullNoTimeout() throws Exception {
-    assertFalse(sut.check("foo"));
-    assertFalse(sut.check(null));
+  public void testTimeout95ms() throws Exception {
+    request = new DummyRequest();
+    Thread.sleep(95);
```

Comment:
a small suggestion: refactor `TimeoutChecker` to use a `java.time.Clock` instance for checking elapsed time instead, which you can then control with a fake/fixed Clock in the unit test, removing the need for `Thread.sleep` in tests.

(I'm also not sure if you can rely on a thread being slept for precisely 95 or 101 milliseconds)

Reply:
I was going to suggest something similar, but figured we could refactor that afterwards

Reply:
I like the idea of injecting a custom Clock into TimeoutChecker and using a FixedClock for testing. the good thing is that TimeoutChecker is an internal class and can be refactored later without breaking public API.

alas, I am not very familiar with Clock API so I cannot provide an immediate good-quality solution using it. my first glance suggests that there may be some issues with **nanoseconds precision** with Clock class.
e.g. this sample code shows very different values returned by `System.nanoTime()` and by `clock.instant.xxx`:
```    
    long createdNanos = System.nanoTime();
    Clock clock = Clock.systemUTC();
    long anotherNanos = TimeUnit.NANOSECONDS.convert(clock.instant().toEpochMilli(), TimeUnit.MILLISECONDS);
    System.out.println(createdNanos + " request nanos");
    System.out.println(anotherNanos + " request nanos - alternative");
```
output: 

> 1885962380543984 request nanos
> 1548885786068000000 request nanos - alternative

Reply:
Good point - I forgot that Clock didn’t provide a similar method to nanoTime(). 👍